src/libraries/System.Data.Common/src/System/Data/Common/DbBatch.cs

```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Data.Common
+{
+    public abstract class DbBatch : IDisposable, IAsyncDisposable
+    {
+        public DbBatchCommandCollection BatchCommands => DbBatchCommands;
+
+        protected abstract DbBatchCommandCollection DbBatchCommands { get; }
+
+        public abstract int Timeout { get; set; }
+
+        public DbConnection? Connection { get; set; }
+
+        protected abstract DbConnection? DbConnection { get; set; }
+
+        public DbTransaction? Transaction { get; set; }
+
+        protected abstract DbTransaction? DbTransaction { get; set; }
+
+        public DbDataReader ExecuteReader()
+            => ExecuteDbDataReader();
+
+        protected abstract DbDataReader ExecuteDbDataReader();
+
+        public Task<DbDataReader> ExecuteReaderAsync(CancellationToken cancellationToken = default)
+            => ExecuteDbDataReaderAsync(cancellationToken);
+
+        protected abstract Task<DbDataReader> ExecuteDbDataReaderAsync(CancellationToken cancellationToken);
+
+        public abstract int ExecuteNonQuery();
+
+        public abstract Task<int> ExecuteNonQueryAsync(CancellationToken cancellationToken = default);
+
+        public abstract object? ExecuteScalar();
+
+        public abstract Task<object?> ExecuteScalarAsync(CancellationToken cancellationToken = default);
+
+        public abstract void Prepare();
+
+        public abstract Task PrepareAsync(CancellationToken cancellationToken = default);
+
+        public abstract void Cancel();
+
+        public DbBatchCommand CreateBatchCommand() => CreateDbBatchCommand();
+
+        protected virtual DbBatchCommand CreateDbBatchCommand() => throw new NotSupportedException();
+
+        public virtual void Dispose() {}
```

Comment:
@stephentoub can you please confirm that this (and the below DisposeAsync) is acceptable in an abstract base class, given there are no resources to be disposed on it, and that [the dispose pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposebool) could simply be implemented by implementations if needed?

Also, I've made this non-abstract since most implementations aren't expected to override.

Reply:
I'm a little surprised to not see the standard dispose pattern implemented here, although I do expect the inheritance hierarchy to be very shallow (a single derived class per provider, and it should be `sealed`), thus I can't imagine it being a problem in practice.

Reply:
@bgrainger missed this originally.

I think the standard dispose pattern makes sense when the base class already have some disposable resource; DbCommand/DbConnection extend Component (DbBatch doesn't), which itself is disposable (mainly for event management, AFAIK), which is why the pattern is followed there. For an empty base class with no disposable resources, I'm not sure what the pattern would bring - it can always be added in implementing classes, if those classes have a disposable resource and expect to be themselves inherited...

Let me know if you see any issue with this logic.